cudax/include/cuda/experimental/__memory_resource/any_resource.h

```diff
@@ -0,0 +1,248 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__MEMORY_RESOURCE_ANY_RESOURCE_H
+#define _CUDAX__MEMORY_RESOURCE_ANY_RESOURCE_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if !defined(_CCCL_COMPILER_MSVC_2017) && defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+
+#  include <cuda/__memory_resource/get_property.h>
+#  include <cuda/__memory_resource/resource.h>
+#  include <cuda/__memory_resource/resource_ref.h>
+#  include <cuda/std/__concepts/_One_of.h>
+#  include <cuda/std/__memory/construct_at.h>
+#  include <cuda/std/__type_traits/is_nothrow_constructible.h>
+#  include <cuda/std/__type_traits/remove_cvref.h>
+#  include <cuda/std/__utility/exchange.h>
+#  include <cuda/std/__utility/forward.h>
+#  include <cuda/std/__utility/in_place.h>
+
+#  if _CCCL_STD_VER >= 2014
+
+namespace cuda::experimental::mr
+{
+template <class _Ty, class _Uy = _CUDA_VSTD::remove_cvref_t<_Ty>>
+_LIBCUDACXX_INLINE_VAR constexpr bool __is_basic_any_resource = false;
+
+template <_CUDA_VMR::_AllocType _Alloc_type, class... _Properties>
+class basic_any_resource
+    : public _CUDA_VMR::_Resource_ref_base<_Alloc_type, false>
+    , private _CUDA_VMR::_Filtered_vtable<_Properties...>
+{
+private:
+  template <_CUDA_VMR::_AllocType, class...>
+  friend class basic_any_resource;
+
+  template <class...>
+  friend struct _CUDA_VMR::_Resource_vtable;
+
+  using __vtable = _CUDA_VMR::_Filtered_vtable<_Properties...>;
+
+  template <class... _OtherProperties>
+  static constexpr bool __properties_match =
+    _CUDA_VSTD::__all_of<_CUDA_VSTD::_One_of<_Properties, _OtherProperties...>...>;
+
+  template <typename _Resource>
+  basic_any_resource(_CUDA_VSTD::in_place_t,
+                     _Resource&& __res) noexcept(_CUDA_VMR::_IsSmall<_CUDA_VSTD::remove_cvref_t<_Resource>>())
+      : _CUDA_VMR::_Resource_ref_base<_Alloc_type, false>(
+          nullptr, &_CUDA_VMR::__alloc_vtable<_Alloc_type, _CUDA_VSTD::remove_cvref_t<_Resource>>)
+      , __vtable(__vtable::template _Create<_CUDA_VSTD::remove_cvref_t<_Resource>>())
```

Comment:
Suggestion: Sometimes when we use a dependent type, we add it to the template header. Might make it more readable
```suggestion
  template <typename _Resource,  typename __resource_t= _CUDA_VSTD::remove_cvref_t<_Resource>>
  basic_any_resource(_CUDA_VSTD::in_place_t, _Resource&& __res) noexcept(_CUDA_VMR::_IsSmall<__resource_t>())
      : _CUDA_VMR::_Resource_ref_base<_Alloc_type, false>(nullptr, &_CUDA_VMR::__alloc_vtable<_Alloc_type, __resource_t>)
      , __vtable(__vtable::template _Create<__resource_t>())
```